tasks/gaia.js

```diff
@@ -0,0 +1,302 @@
+    `${cliBinary} tx create-validator` +
+    ` --home ${clientHomeDir}` +
+    ` --from ${keyName}` +
+    ` --amount=10steak` +
+    ` --pubkey=${valPubKey}` +
+    ` --address-delegator=${operatorAddress}` +
+    ` --moniker=${moniker}` +
+    ` --chain-id=test_chain` +
+    ` --commission-max-change-rate=0` +
+    ` --commission-max-rate=0` +
+    ` --commission-rate=0` +
+    ` --json`
+
+  return makeExecWithInputs(command, [password])
+}
+
+async function sendTokens(
+  { keyName, password, clientHomeDir }, // senderSignInfo
+  tokenString, // like "10stake" <- amount followed by denomination
+  toAddress
+) {
+  let command =
+    `${cliBinary} tx send` +
+    ` --home ${clientHomeDir}` +
+    ` --from ${keyName}` +
+    ` --amount=${tokenString}` +
+    ` --to=${toAddress}` +
+    ` --chain-id=test_chain`
+  return makeExecWithInputs(command, [password], false)
+}
+
+// start a node and connect it to nodeOne
+// nodeOne is used as a persistent peer for all the other nodes
+// wait for blocks to show as a proof, the node is running correctly
+function startLocalNode(
+  nodeHome,
+  number = 1, // number is used to prevent conflicting ports when running multiple nodes
+  nodeOneId = ``
+) {
+  return new Promise((resolve, reject) => {
+    let command = `${nodeBinary} start --home ${nodeHome}`
+    if (number > 1) {
+      // setup different ports
+      command += ` --p2p.laddr=tcp://0.0.0.0:${defaultStartPort -
+        (number - 1) * 3} --address=tcp://0.0.0.0:${defaultStartPort -
+        (number - 1) * 3 +
+        1} --rpc.laddr=tcp://0.0.0.0:${defaultStartPort - (number - 1) * 3 + 2}`
+      // set the first node as a persistent peer
+      command += ` --p2p.persistent_peers="${nodeOneId}@localhost:${defaultStartPort}"`
+    }
+    console.log(command)
+    const localnodeProcess = spawn(command, { shell: true })
+
+    // log output for debugging
+    const logPath = path.join(nodeHome, `process.log`)
+    console.log(`Redirecting node ` + number + ` output to ` + logPath)
+    fs.createFileSync(logPath)
+    let logStream = fs.createWriteStream(logPath, { flags: `a` })
+    localnodeProcess.stdout.pipe(logStream)
+
+    localnodeProcess.stderr.pipe(process.stderr)
+
+    // wait 20s for the first block or assume the node has failed
+    const timeout = setTimeout(() => {
+      reject(`Timed out waiting for block for node ${number}`)
+    }, 20000)
+
+    // wait for a message about a block being produced
+    function listener(data) {
+      let msg = data.toString()
+
+      if (msg.includes(`Block{`)) {
+        localnodeProcess.stdout.removeListener(`data`, listener)
+        console.log(`Node ` + number + ` is running`)
+        clearTimeout(timeout)
+        resolve()
+      }
+    }
+
+    console.log(`Waiting for first block on node ` + number)
+    localnodeProcess.stdout.on(`data`, listener)
+
+    localnodeProcess.once(`exit`, reject)
+  })
+}
+
+// execute command and return stdout
+function makeExec(command) {
```

Comment:
If you use `const makeExec = command => ` instead then it won't get hoisted and I would have spent less time looking for it.

Reply:
I like hoisting with helper functions. Keeps the top of the file for the important functions. But I understand your approach.